Enable Search With lunr.js

hamelsmu opened the pull request:
[Lunr.js](https://lunrjs.com/) allows us to power a client-side search.  

- No ads from google
- Runs completely on the client side
- Allows your users to quickly search for blog articles based on keyword search

![gif-search-small](https://user-images.githubusercontent.com/1483922/75119836-6a051a80-563b-11ea-8818-d569cecdeba7.gif)






@jph00 what do you think?  If this is too much feature bloat, let me know as well.  Maybe getting too excited / carried away.  Edit: see the preview link to preview the search preview

hamelsmu commented:
/preview

hamelsmu commented:
@jph00 incase you are reading this via email the search can be previewed via this link: https://5e52e678fa14ebe98cc91602--fastpages.netlify.com/search/

jph00 commented:
It looks amazing. @sgugger I suspect we should use this in nbdev (not urgent though) - what do you think?

@hamelsmu is the index built automatically on push?

xnutsive commented on _config.yml:
```diff
@@ -40,6 +40,7 @@ exclude:
   - settings.ini
   - docker-compose.yml
   - action.yml
+  - README_TEMPLATE.md
```
I wasn't sure if this should be ignored — thank you for adding it!

hamelsmu commented:
Yes the index is built on `jekyll build` so it will happen automatically on push

xnutsive commented on _includes/search-lunr.html:
```diff
@@ -0,0 +1,71 @@
+<script src="/js/lunr.js"></script>
```
If we use the js file from our repository, what would be the best way to manage versions and updates to future versions of this and other js libs we'll use?

hamelsmu replied:
to be honest, I have no idea.  If we vendor the script in the repo, do we ever need to change it?  (Sorry if this is a dumb question)

hamelsmu replied:
Sorry let me re-phrase my question whats the best alternative?  What other options do we have?

xnutsive commented on _includes/search-lunr.html:
```diff
@@ -0,0 +1,71 @@
+<script src="/js/lunr.js"></script>
+
+<script>
```
I'll be the bad cop and ask if it makes sense to create a separate js file, say àssets/lunr-search.js` that would handle the index? However if this is only used once, and the whole file is small and readable, it's fine as it is, too.

hamelsmu replied:
Sure I'll move it to `/assets/js/vendor`

xnutsive commented on _includes/search-lunr.html:
```diff
@@ -0,0 +1,71 @@
+
+<script>
+{% assign counter = 0 %}
+var documents = [{% for page in site.pages %}{% if page.url contains '.xml' or page.url contains 'assets' or page.search == 'exclude' %}{% else %}{
```
Since the index is client-side, this will put all the pages with all their content into this JS snippet. This will become a pretty big chunk of text for larger blogs pretty quickly. 

It should work great as long as the search lives on a separate page. If we'd want to have a search bar in the header that searches the blog, on any page — that would mean that we'll download the full blog on every page.

I don't know if there's a good workaround for this — it's definitely good enough for now, until we see any problems and issues from our users. We could go ahead and use Algolia instead, but that's a 3rd party dependency I'd rather not have, or have as an opt-in.

hamelsmu replied:
That's good to know, I guess we will have search on a separate page for now

xnutsive commented on _includes/search-lunr.html:
```diff
@@ -0,0 +1,71 @@
+    return false;
+}
+</script>
+<style>
```
Should we move styles to àssets/main.scss`?

hamelsmu replied:
makes sense

xnutsive commented on search.html:
```diff
@@ -0,0 +1,8 @@
+---
```
If `search.html` is the only place where we use the search form, would it make sense to just put the whole page in this file, and not have the include snippet at all?

xnutsive commented on _includes/search-lunr.html:
```diff
@@ -0,0 +1,71 @@
+                var ref = results[i]['ref'];
+                var url = documents[ref]['url'];
+                var title = documents[ref]['title'];
+                var body = documents[ref]['body'].substring(0,160)+'...';
```
I've looked for a way to show not just the beginning of a matching article, but a specific excerpts that matched the query. 

It seems there's a neat way to do this: 

1. Allow indexer to store term positions metadata: 

```js
var idx = lunr(function () {
  this.ref('id')
  this.field('body')
  this.metadataWhitelist = ['position']

  documents.forEach(function (doc) { this.add(doc) }, this)
})
```

2. Then, search results will have term positions in the body of the document:

```json
{
  "ref": "123",
  "score": 0.123456,
  "metadata": {
    "test": {
      "body": {
        "position": [[0, 4], [24, 4]]
      }
    }
  }
}
```

With that, we can show what exactly has been matched! Seeing how quickly fastpages improves, I'd say merge without excerpts highlighting, and add them later on. This will take some time to implement correctly, like: 

- What if a matched term is in a code block? We'll need to do code highlighting in search results then. 
- But if we do code blocks on search results, we can't just cut them with `substring(0,160)`. We'll need to do variable length smart text size limiting for search results. 
- Also, if we'll support code search results, and basically any other HTML as well, we'll have to match content from markdownified, indexed version, back to the HTML original. 
- And finally, we'll need to write styling and HTML around the highlighted terms (yellow background with Primer, etc). 

What do you think?

hamelsmu replied:
I think this is a great idea.  I have tried really hard to do this myself or find a library that supports this out of the box, and I have found some tutorials but they have been really hard for me to follow.  Maybe you would have better luck at attempting this vs. me 🙇

hamelsmu commented:
@xnutsive thanks for the feedback.  I am a total js/html/ruby etc. noob so I will give it my best shot and see if I can move things around etc.  I will admit I found this solution on the internet and just mindlessly followed the instructions :), so bear with me as a fumble through this - good suggestions

hamelsmu commented:
@xnutsive actually have another question

I was looking at this Jekyll site called [Just The Docs](https://pmarsceill.github.io/just-the-docs/), and they also build their search on lunr.js.  However their search is really, really fancy looking with keyword highlighting etc. 

Here is the [Just The Docs repo](https://github.com/pmarsceill/just-the-docs).  I have been too dumb to try to replicate their functionality despite hacking on it for about an hour, perhaps you can take a look and let me know if you see an easy path to borrow what they have?  

Thanks for your help

hamelsmu commented:
### Summary of changes made

- moved lunr js file into `/assets/js/vendor`
- bumped version of lunr from 2.1 to 2.3 and using a minified file instead
- moved css from search page into main css file
- removed seperate includes page for search since there is only one search page

/preview

hamelsmu commented:
/preview

hamelsmu commented:
I made all the requested changes since everyone has approved this PR I'll go ahead and merge.

xnutsive commented:
Almost missed it! Thanks for the link — there are a few interesting things with Just the Docs, I'll go ahead and do another issue with those.
